lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs

```diff
@@ -0,0 +1,129 @@
+            { new TaskStateTransition(TaskTransitionState.TaskWaitingForClose, TaskEvent.ClosedTask), TaskTransitionState.TaskClosedByDriver },
+            { new TaskStateTransition(TaskTransitionState.TaskWaitingForClose, TaskEvent.FailedTaskAppError), TaskTransitionState.TaskFailedAppError },
+            { new TaskStateTransition(TaskTransitionState.TaskWaitingForClose, TaskEvent.FailedTaskSystemError), TaskTransitionState.TaskFailedSystemError },
+            { new TaskStateTransition(TaskTransitionState.TaskWaitingForClose, TaskEvent.FailedTaskEvaluatorError), TaskTransitionState.TaskFailedByEvaluatorFailure },
+            { new TaskStateTransition(TaskTransitionState.TaskWaitingForClose, TaskEvent.FailedTaskCommunicationError), TaskTransitionState.TaskFailedByGroupCommunication },
+            { new TaskStateTransition(TaskTransitionState.TaskFailedSystemError, TaskEvent.FailedTaskEvaluatorError), TaskTransitionState.TaskFailedByEvaluatorFailure },
+            { new TaskStateTransition(TaskTransitionState.TaskFailedByGroupCommunication, TaskEvent.FailedTaskEvaluatorError), TaskTransitionState.TaskFailedByEvaluatorFailure }
+        };
+
+        private static TaskTransitionState[] FinalStatae = 
+        {
+            TaskTransitionState.TaskFailedAppError,
+            TaskTransitionState.TaskFailedSystemError,
+            TaskTransitionState.TaskFailedByEvaluatorFailure,
+            TaskTransitionState.TaskFailedByGroupCommunication,
+            TaskTransitionState.TaskClosedByDriver,
+            TaskTransitionState.TaskCompleted
+        };
+
+        private volatile TaskTransitionState _currentState;
+        private readonly object _lock = new object();
+
+        /// <summary>
+        /// Create a new DriverTaskState with TaskNew as the task initial state
+        /// </summary>
+        public DriverTaskState()
+        {
+            _currentState = TaskTransitionState.TaskNew;
+        }
+
+        /// <summary>
+        /// return the current task state
+        /// </summary>
+        public TaskTransitionState CurrentState
+        {
+            get
+            {
+                return _currentState;
+            }
+        }
+
+        private TaskTransitionState GetNext(TaskEvent taskEvent)
+        {
+            TaskStateTransition transition = new TaskStateTransition(_currentState, taskEvent);
+            TaskTransitionState nextState;
+            if (!Transitions.TryGetValue(transition, out nextState))
+            {
+                throw new ApplicationException("Invalid transition: " + _currentState + " -> " + taskEvent);
+            }
+            return nextState;
+        }
+
+        /// <summary>
+        /// Move to the next state
+        /// </summary>
+        /// <param name="command"></param>
+        /// <returns></returns>
+        public TaskTransitionState MoveNext(TaskEvent command)
+        {
+            lock (_lock)
+            {
+                _currentState = GetNext(command);
```

Comment:
Depends. Do we want to be able to ask `what-if` on this object? `GetNext()` allows that without changing state of the object.